### DISCUSSION TITLE
How to perform an ECS Service replace action without outage

### DISCUSSION THREAD

**Discussion by nicodeg87** [pulumi/pulumi#9389] [Pulumi in Practice] [answered]
Hi guys,

We are trying to setup pulumi to deploy our ECS services and we realised about an edge case  which is when we need to perform an update over an ECS service that triggers a replace action (for example changing the capacity provider). In this scenario, pulumi is creating a new service with a new name with the random suffix, BUT the problem is that it removes the previous service before the tasks in the new service are on running state and as consequence it produces downtime on the service.

I tried ti look for a custom option  to manipulate the time before the deletion of a resource on a replace action, but I could not find any configuration related to this. Is there any option to handle this scenario? 

Thanks,
Nicolas.

**Comment by jaxxstorm** [comment, marked as answer]
Hi @nicodeg87 !

This is really an ECS problem. Pulumi won't start removing the old ECS service until the new one has fully provisioned, however it sounds like in your case, the ECS services outputs have all resolved, even though the application in the task definition isn't healthy.

The way to fix this is to use some properties on the ECS service definition:

Firstly, make sure the healthcheck on your load balancer is accurate and reporting correctly: https://www.pulumi.com/registry/packages/aws/api-docs/alb/targetgroup/#health_check_python

Then, make sure the deployment minimum healthy percent is at a reasonable level: https://www.pulumi.com/registry/packages/aws/api-docs/ecs/service/#deployment_minimum_healthy_percent_python



There's more information here: https://aws.amazon.com/blogs/compute/bluegreen-deployments-with-amazon-ecs/

**Comment by nicodeg87** [comment]
Hi @jaxxstorm!

Thanks a lot for the quick response. Answering your comments:

> Firstly, make sure the healthcheck on your load balancer is accurate and reporting correctly:

We are not using load balancer as we use our own gateway/service discovery solution.

> Then, make sure the deployment minimum healthy percent is at a reasonable level: 

We have it set at 100%. So it should not be the reason for this behavior.

> There's more information here: https://aws.amazon.com/blogs/compute/bluegreen-deployments-with-amazon-ecs/

We are using rolling update ECS deployment strategy with circuit breaker enabled.

Could it be related to the usage of [waitForSteadyState](https://www.pulumi.com/registry/packages/aws/api-docs/ecs/service/#waitforsteadystate_nodejs) ?

**Comment by nicodeg87** [comment]
I just checked with waitForSteadyState and it worked good. The service was removed after the new service was on running state.

Thanks for the support!
